src/timemachine-heap.c

```diff
@@ -0,0 +1,150 @@
+    if (unlikely(heap == NULL)) {
+        return NULL;     
+    }
+
+    heap->conf = conf;
+    heap->used_pool_count = 0;
+    heap->unused_pool_count = 0;
+    
+    TAILQ_INIT(&heap->used_mem_pools);
+    TAILQ_INIT(&heap->unused_mem_pools);
+    TAILQ_INIT(&heap->unused_packets);    
+    
+    uint32_t expanded = TimeMachineHeapExpand(tmtv, heap, conf->prealloc_count);
+    if (expanded != conf->prealloc_count) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Error: Could not allocate initial memory for "
+                   "time machine, expanded heap by: %" PRIu32 ", requested: %" PRIu32 
+                   " current Memory: %" PRIu64 " max Memory: %" PRIu64, expanded, 
+                   conf->prealloc_count, tmtv->current_memory, timemachine_config.max_memory);
+        exit(EXIT_FAILURE);                         
+    }
+    
+    return heap;
+}
+
+void TimeMachineHeapDestroy(TimeMachineHeap* heap) {
+    
+    if (heap == NULL) {
+        return;
+    }
+    
+    /* delete all the used pool data */
+    while (heap->used_pool_count > 0) {
+        TimeMachineMemPool* mem_pool;
+            
+        mem_pool = TAILQ_FIRST(&heap->used_mem_pools);
+        TAILQ_REMOVE(&heap->used_mem_pools, mem_pool, next);
+        SCFree(mem_pool);
+
+        heap->used_pool_count--;    
+    }
+    
+    /* delete all the unused pool data */
+    while (heap->unused_pool_count > 0) {
+        TimeMachinePacket* packet;
+        TimeMachineMemPool* mem_pool;
+
+        packet = TAILQ_FIRST(&heap->unused_packets);
+        TAILQ_REMOVE(&heap->unused_packets, packet, next);
+        TimeMachinePacketDestroy(packet);
+            
+        mem_pool = TAILQ_FIRST(&heap->unused_mem_pools);
+        TAILQ_REMOVE(&heap->unused_mem_pools, mem_pool, next);
+        SCFree(mem_pool);
+                
+        heap->unused_pool_count--;
+    }    
+}
+
+uint32_t TimeMachineHeapExpand(TimeMachineThreadVars* tmtv, TimeMachineHeap* heap, uint32_t expand_by) {
+    uint32_t i;
+           
+    if (heap == NULL) {
+        return 0;
+    }
+
+    uint32_t expand_count = expand_by;
+    uint32_t expand_mem_size = sizeof(TimeMachineMemPool) + heap->conf->max_packet_size + 
+                      sizeof(TimeMachinePacket);
+                   
+    if ((expand_mem_size * expand_by) + tmtv->current_memory > timemachine_config.max_memory) {
+        expand_count = (timemachine_config.max_memory - tmtv->current_memory) / expand_mem_size + 1;
+    }    
+    
+    for (i = 0; i < expand_count; i++) {
+        TimeMachineMemPool* mem_pool;
+        TimeMachinePacket* packet;
+        
+        mem_pool = SCCalloc(sizeof(TimeMachineMemPool), 1);
+        if (mem_pool == NULL) {
+            return i;
+        }    
+        
+        mem_pool->mem = SCMalloc(heap->conf->max_packet_size);
+        if (mem_pool->mem == NULL) {
+            SCFree(mem_pool);
+            return i;
+        }
+        
+        packet = TimeMachinePacketNew();
```

Comment:
I think I see where the confusion lies.  The time machine heaps are created per thread, not per flow.  The heaps are just the pre-allocated memory pools to store the packet data.  TimeMachinePacket is a data structure that ensures packets across multiple heaps are associated to the same flow.  TimeMachinePackets don't actually contain data, just point to the memory pools across the given heaps that do.  Having a heaps allocated at the thread level makes resource management easier while having the logical ordering of the packets at the flow level is an optimization.  Otherwise, you would need to iterate through the entire contents of each heap which depending on size, could range into the millions of packets per heap, if not larger (a 1 GB heap with a packet size of 200 bytes would be around 5+ million packets).  

When a mempool is removed from the heap, lets say because the size of the heap is too large, then the packet also needs to be removed from the flow.  What this does mean is that there is a case where you may not have all packets associated with a given flow.  If one of your heaps is too busy, data may age off before an alarm fires so you would have a missing hole in your logical ordering of packets at the flow level.  In generally this shouldn't be a significant issue as the majority of alarms that fire are primarily geared towards the beginning of a session.